src/utils.rs

```diff
@@ -49,23 +49,21 @@ pub(crate) fn pluralize<'a>(count: usize, singular: &'a str, plural: &'a str) ->
 }
 
 /// Splits the input text into distinct trees,
-/// delimited by two successive newlines
-///
-/// This function is called before the tokenization and parsing steps.
+/// delimited by two successive newlines.
 #[inline]
-pub(crate) fn split_trees(text: &str) -> Vec<&str> {
-    text.split(TREES_SEPARATOR).collect::<Vec<&str>>()
+pub(crate) fn split_trees(text: &str) -> impl Iterator<Item = &str> + '_ {
+    text.split(TREES_SEPARATOR)
 }
 
 /// Gets the contract name from the HIR tree identifier.
-///
-/// This function is called during verification of the combined HIR.
-pub(crate) fn get_contract_name_from_identifier(identifier: &String) -> String {
-    let identifier_parts: Vec<&str> = identifier.split(CONTRACT_PART_SEPARATOR).collect();
-    return sanitize(identifier_parts[0]);
+pub(crate) fn get_contract_name_from_identifier(identifier: &str) -> String {
+    let contract_name = identifier
+        .split(CONTRACT_PART_SEPARATOR)
+        .next()
+        .expect("should not be empty");
```

Comment:
@alexfertel if this expectation is violated, will this cause an unhandled panic? If so, does this need to be propagated to the calling function?

Reply:
Ah, very good catch. That should actually be something like:

```rust
pub(crate) fn get_contract_name_from_identifier(identifier: &str) -> Option<String> {
    identifier
        .split(CONTRACT_PART_SEPARATOR)
        .next()
        .map(sanitize)
}
```

This is better because it reflects that the contract name might be malformed. Then at the call site, it would be:

```rust
                    let identifier = get_contract_name_from_identifier(&text).unwrap_or_default();
```
Since we handle the id being different a couple of lines below that one.

---

The problem is how do we handle the first tree having an empty tree. This leads me to think that we should fail on a previous compiler pass if we don't find the contract name.

So, to summarize, we should:
- Error if a tree starts with `::`, i.e. `::a_function`.
- Make `get_contract_name_from_identifier` panic if this assumption is not held, because that is a bug in `bulloak`

Does that make sense?

---

> Additionally, both check and scaffold are currently erroring with "a Corner must be the last child" – I'd thought this wouldn't be an issue since we are splitting the trees before forwarding on to the parser, but for some reason it is, and I haven't yet been able to track down the cause. If you have any ideas that would be great:)

Yeah, your mental model is correct, I'm unsure why this happens. I'll have a look tomorrow and see what that's about.

Reply:
Update: I am currently unable to reproduce what you found.

I added a `multiple_roots.tree` file with the following contents:

```
Contract::function1
├── It should never revert.
└── When first arg is bigger than second arg
    └── It is all good

Contract::function2
└── When stuff happens
    └── It should do something simple
```

And it correctly generates the following contract:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.0;

contract Contract {
    function test_ShouldNeverRevert() external {
        // It should never revert.
    }

    function test_WhenFirstArgIsBiggerThanSecondArg() external {
        // It is all good
    }

    function test_WhenStuffHappens() external {
        // It should do something simple
    }
}
```

Maybe what's happening is that the last child on the tree you are testing with is not a `Corner` (`└`)?

Reply:
Great, yeah that makes sense.

> Update: I am currently unable to reproduce what you found.

Weird. I'm still having this issue with the tree you provided, so quite probably a problem on my end. Having said that, I haven't really changed anything, so if hard reset doesn't work, then I'm stumped. `cargo build --release` should build with the latest changes, right?

Reply:
Hmmm, you can clean your local setup with `cargo clean`, and then what happens if you do `cargo run -- scaffold tests/scaffold/multiple_roots.tree`?

Also, maybe you have commits that haven't been pushed?